homeassistant/components/pvpc_hourly_pricing/sensor.py

```diff
@@ -0,0 +1,394 @@
+from homeassistant.const import ATTR_ATTRIBUTION, CONF_NAME, CONF_TIMEOUT
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.entity import async_generate_entity_id
+from homeassistant.helpers.event import (
+    async_track_point_in_time,
+    async_track_time_change,
+)
+from homeassistant.helpers.restore_state import RestoreEntity
+import homeassistant.util.dt as dt_util
+
+from . import SENSOR_SCHEMA
+from .const import ATTR_TARIFF, DEFAULT_TIMEOUT, DOMAIN, TARIFFS
+
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(SENSOR_SCHEMA.schema)
+
+_ATTRIBUTION = "Data retrieved from api.esios.ree.es by REE"
+_PRECISION = 5
+_RESOURCE = (
+    "https://api.esios.ree.es/archives/70/download_json"
+    "?locale=es&date={day:%Y-%m-%d}"
+)
+_TARIFF_KEYS = dict(zip(TARIFFS, ["GEN", "NOC", "VHC"]))
+
+# Prices are given in 0 to 24h sets, adjusted to the main timezone in Spain
+_REFERENCE_TZ = timezone("Europe/Madrid")
+
+ATTR_PRICE = "price"
+ICON = "mdi:currency-eur"
+UNIT = "€/kWh"
+
+
+async def async_setup_platform(
+    hass: HomeAssistant, config, async_add_devices, discovery_info=None
+):
+    """
+    Set up the electricity price sensor as a sensor platform.
+
+    ```yaml
+    sensor:
+      - platform: pvpc_hourly_pricing
+        name: pvpc_manual_sensor
+        tariff: normal
+
+      - platform: pvpc_hourly_pricing
+        name: pvpc_manual_sensor_2
+        tariff: discrimination
+        timeout: 8
+    ```
+    """
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN, data=config, context={"source": config_entries.SOURCE_IMPORT}
+        )
+    )
+    return True
+
+
+async def update_listener(hass: HomeAssistant, entry: config_entries.ConfigEntry):
+    """Update selected tariff for sensor."""
+    if entry.options[ATTR_TARIFF] != entry.data[ATTR_TARIFF]:
+        entry.data[ATTR_TARIFF] = entry.options[ATTR_TARIFF]
+        hass.async_create_task(hass.config_entries.async_reload(entry.entry_id))
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, config_entry: config_entries.ConfigEntry, async_add_entities
+):
+    """Set up the electricity price sensor from config_entry."""
+    if not config_entry.update_listeners:
+        config_entry.add_update_listener(update_listener)
+
+    name = config_entry.data[CONF_NAME]
+    entity_id = async_generate_entity_id(ENTITY_ID_FORMAT, name, hass=hass)
+    async_add_entities(
+        [
+            ElecPriceSensor(
+                websession=async_get_clientsession(hass),
+                name=name,
+                entity_id=entity_id,
+                tariff=config_entry.data[ATTR_TARIFF],
+                timeout=config_entry.data.get(CONF_TIMEOUT, DEFAULT_TIMEOUT),
+            )
+        ],
+        True,
+    )
+
+
+def extract_prices_for_tariff(data: List[dict], tariff: str) -> Dict[datetime, float]:
```

Comment:
Please move this processing into the pypi package

Reply:
Done